flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala

```diff
@@ -488,19 +499,50 @@ abstract class TableEnvironment(val config: TableConfig) {
     *   tEnv.sql(s"SELECT * FROM $table")
     * }}}
     *
-    * @param query The SQL query to evaluate.
-    * @return The result of the query as Table.
+    * @param sql The SQL string to evaluate.
+    * @return The result of the query as Table or null of the DML insert operation.
     */
-  def sql(query: String): Table = {
+  def sql(sql: String): Table = {
     val planner = new FlinkPlannerImpl(getFrameworkConfig, getPlanner, getTypeFactory)
     // parse the sql query
-    val parsed = planner.parse(query)
+    val parsed = planner.parse(sql)
     // validate the sql query
     val validated = planner.validate(parsed)
-    // transform to a relational tree
-    val relational = planner.rel(validated)
 
-    new Table(this, LogicalRelNode(relational.rel))
+    // separate source translation if DML insert, considering sink operation has no output but all
+    // RelNodes' translation should return a DataSet/DataStream, and Calcite's TableModify
+    // defines output for DML(INSERT,DELETE,UPDATE) is a RowCount column of BigInt type. This a
+    // work around since we have no conclusion for the output type definition of sink operation
+    // (compare to traditional SQL insert).
+    if (parsed.isInstanceOf[SqlInsert]) {
+      val insert = parsed.asInstanceOf[SqlInsert]
```

Comment:
The `isInstanceOf` / `asInstanceOf` can be more elegantly implemented with pattern matching.